# QC and Exploratory Analysis


!!! r-project "code"

    ```r
    library(DropletUtils)
    library(scater)
    library(ensembldb)
    library(AnnotationHub)
    library(BiocParallel)
    library(tidyverse)
    library(patchwork)
    library(ggvenn)
    ```

!!! r-project "code"

    ```r
    setwd("/nesi/project/nesi02659/sc-rna")
    samplesheet <- read.csv("Data/sample_sheet.tsv")
    ```

    - Here were are selecting to use forked processes with MulticoreParam and instructing the function to use 4 cores 
    ```r
    bp.params <- MulticoreParam(workers = 4)
    ```
    ```r
    sample.path <- "Data/CellRanger_Outputs/SRR9264343/outs/filtered_feature_bc_matrix/"
    sce.sing <- read10xCounts(sample.path, col.names=TRUE, BPPARAM = bp.params)
    sce.sing
    ```
    ??? circle-check "Output"

        ```bash
        class: SingleCellExperiment 
        dim: 37764 3153 
        metadata(1): Samples
        assays(1): counts
        rownames(37764): ENSG00000243485 ENSG00000237613 ...
          ENSG00000278817 ENSG00000277196
        rowData names(3): ID Symbol Type
        colnames(3153): AAACCTGAGACTTTCG-1 AAACCTGGTCTTCAAG-1 ...
          TTTGTCACAGGCTCAC-1 TTTGTCAGTTCGGCAC-1
        colData names(2): Sample Barcode
        reducedDimNames(0):
        mainExpName: NULL
        altExpNames(0):
        ```
### The SingleCellExperiment object

!!! quote ""

    The data have been loaded as a [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) object. The details of the structure of the object are described here. In summary, it stores various data types in a single object. Currently it will contain:
    
    - the count matrix
    - feature (gene) metadata
    - cell (droplet) metadata
    
    Later we will also add the outcomes of downstream analysis such as normalisation and dimensionality reduction.
    
    ![image](../r_images/single_cell_experiment.png)



### The counts matrix 

Compared to bulk RNA-seq, Single-cell RNA-seq data is sparse, i.e. there many missing values or zeroes. This is particularly true with droplet-based methods such as 10X, mostly because:

- any given cell does not express each gene
- the library preparation does not capture all transcript the cell does express
- the sequencing depth per cell is far lower and so fewer of the expressed genes are detected

We can access the counts matrix with counts. Given the large number of droplets in a sample, count matrices can be large.

!!! r-project "code"

    ```r
    dim(counts(sce.sing))
    ```

    - They are however very sparse, that is, most of the entries are 0’s. To save memory the counts can be stored in a ‘sparse matrix’ that only stores non-zero values, in this case as a dgCMatrix object.
    ```r
    counts(sce.sing)[1:10, 1:10]
    ```
    ??? circle-check "Output"
        ```r
        10 x 10 sparse Matrix of class "dgCMatrix"
          [[ suppressing 10 column names 'AAACCTGAGACTTTCG-1', 'AAACCTGGTCTTCAAG-1', 'AAACCTGGTGCAACTT-1' ... ]]
                                           
        ENSG00000243485 . . . . . . . . . .
        ENSG00000237613 . . . . . . . . . .
        ENSG00000186092 . . . . . . . . . .
        ENSG00000238009 . . . . . . . . . .
        ENSG00000239945 . . . . . . . . . .
        ENSG00000239906 . . . . . . . . . .
        ENSG00000241860 . . . . . . . . . .
        ENSG00000241599 . . . . . . . . . .
        ENSG00000286448 . . . . . . . . . .
        ENSG00000236601 . . . . . . . . . .
        ```

### Features

Details about the “features” (in this case genes) can by accessed using the rowData function. Currently it contains the ensembl gene ID and the gene symbol, which have been derived from the 10x reference used by CellRanger. It also contains a “Type” column, which tells us what sort of data we are looking at; in this case it is “Gene Expression”. If we wish to, we can add further annotation to the features by adding extra columns to this data frame.

!!! r-project "code"

    ```r
    rowData(sce.sing)
    ```
    ??? circle-check "Output - The rows of this table correspond to the rows of the count matrix; the row names of this table will match the row names of the counts matrix - currently these are the Ensembl IDS:"

        ```r
        DataFrame with 37764 rows and 3 columns
                                     ID          Symbol            Type
                            <character>     <character>     <character>
        ENSG00000243485 ENSG00000243485     MIR1302-2HG Gene Expression
        ENSG00000237613 ENSG00000237613         FAM138A Gene Expression
        ENSG00000186092 ENSG00000186092           OR4F5 Gene Expression
        ENSG00000238009 ENSG00000238009 ENSG00000238009 Gene Expression
        ENSG00000239945 ENSG00000239945 ENSG00000239945 Gene Expression
        ...                         ...             ...             ...
        ENSG00000277836 ENSG00000277836 ENSG00000277836 Gene Expression
        ENSG00000278633 ENSG00000278633 ENSG00000278633 Gene Expression
        ENSG00000276017 ENSG00000276017 ENSG00000276017 Gene Expression
        ENSG00000278817 ENSG00000278817 ENSG00000278817 Gene Expression
        ENSG00000277196 ENSG00000277196 ENSG00000277196 Gene Expression
        ```
    ```r
    rownames(counts(sce.sing))[1:6]
    ```
### Droplet annotation

!!! r-project "code"

    ```r
    colData(sce.sing)
    colnames(counts(sce.sing))[1:6]
    ```

## Properties of scRNA-seq data

### Number of genes detected per cell

The number and identity of genes detected in a cell varies greatly across cells: the total number of genes detected across all cells is far larger than the number of genes detected in each cell.

For the current set of samples the total number of genes detected across cells was 26468 out of 37764 gene in the reference, but if we look at the number of genes detected in each cell, we can see that this ranges from 40 to 5547, with a median of 2562.


!!! r-project "code"
    ```r
    genesPerCell <- colSums(counts(sce.sing) > 0)
    plot(density(genesPerCell), main="", xlab="Genes per cell")
    ```
    ![image ](../r_images/1_genespercell.png){width="600"}

###  Total UMI for a gene versus the number of times detected

!!! r-project "code"

    ```r
    plot(rowSums(counts(sce.sing)) / rowSums(counts(sce.sing) > 0),
         rowMeans(counts(sce.sing) > 0),
         log = "x",
         xlab="Mean UMIs per cell",
         ylab="proportion of cells expressing the gene"
    )
    ```
    ![image](../r_images/2_total_umi.png){width="600"}

### Distribution of counts for a gene across cells

!!! r-project "code"

    ```r
    rel_expression <- t( t(counts(sce.sing)) / colSums(counts(sce.sing))) * 100
    rownames(rel_expression) <- rowData(sce.sing)$Symbol
    most_expressed <- sort(rowSums( rel_expression ), decreasing = T)[20:1]
    plot_data <- as.matrix(t(rel_expression[names(most_expressed),]))
    
    boxplot(plot_data, cex=0.1, las=1, xlab="% total count per cell", horizontal=TRUE)
    ```

    ??? circle-check "Output"
    
        ![image](../r_images/3_distribution_of_counts.png)

## Quality Control 

### Load multiple samples