# 4. sctransform: Variant Stabilising Transformation

With scaling normalisation a correlation remains between the mean and variation of expression (heteroskedasticity). This affects downstream dimensionality reduction as the few main new dimensions are usually correlated with library size. `sctransform` addresses the issue by regressing library size out of raw counts and providing residuals to use as normalized and variance-stabilized expression values in some downstream analyses, such as dimensionality reduction.

!!! image ""
    <center>
    ![image](../r_images/raw_vs_normalised.png)
    <small>Effect of scaling normalisation</small>
    </center>

The `sctransform` package is from the Seurat suite of scRNAseq analysis packages. Rather than convert our Single Cell Experiment object into a Seurat object and use the Seurat package’s command `SCTransform`, we will extract the counts matrix from our SCE object and run the variance stabilising transformation (VST) algorithm, using the `sctranform` package’s vst command, directly on the matrix. We can extract the counts matrix - as a dgCMatrix object sparse matrix - using the `counts function

!!! r-project "code"

    ```r
    counts <- counts(sce)
    class(counts)
    ```

## Rationale

In order to demonstrate the rationale behind the using the variance stabilising transformation, we will visually inspect various properties of our data. Our main interest is in the general trends not in individual outliers. Neither genes nor cells that stand out are important at this step; we focus on the global trends.

### Derive gene and cell attributes from the UMI matrix

#### Gene attributes

Gene attributes include for each gene:

- mean UMI count across cells
- number of cells where the gene is detected
- variance of UMI counts across cells
- the mean and variance above on the log10 scale 

!!! r-project "code"

    ```r
    gene_attr <- data.frame(mean = rowMeans(counts), 
                            detection_rate = rowMeans(counts > 0),
                            var = rowVars(counts)) %>% 
      mutate(log_mean = log10(mean)) %>% 
      mutate(log_var = log10(var))
    ```
    ```r
    dim(gene_attr)
    ```
    ```r
    head(cell_attr)
    ```
#### Cell attributes

Attributes include for each cell:

- total UMI count across genes (library size)
- number of genes detected (with at least 1 UMI)

!!! r-project "code"

    ```r
    cell_attr <- data.frame(n_umi = colSums(counts),
                            n_gene = colSums(counts > 0))
    ```
    ```r
    dim(cell_attr)
    ```
    ```r
    head(cell_attr)
    ```
### Mean-variance relationship

For the genes, on the log10 scale we can see that up to a mean UMI count of 0 the variance follows the line through the origin with slope one, i.e. variance and mean are roughly equal as expected under a Poisson model. However, genes with a higher average UMI count show overdispersion compared to Poisson.

!!! r-project "code"

    ```r
    ggplot(gene_attr, aes(log_mean, log_var)) + 
      geom_point(alpha=0.3, shape=16) +
      geom_abline(intercept = 0, slope = 1, color='red')
    ```

    ![image](../r_images/22-sctransform-meanvariancership.png)
