# 8. Identification of cluster marker genes

In order to aid the interpretation of the clustering results that we covered in the previous section, it is helpful to identify genes that contribute to the separation of cells into those clusters.

The main approach to achieve this, is to identify genes that are differently expressed between clusters. These may be, for example, exclusively expressed in a single cluster or perhaps differentiate between a few different clusters. There are different methods to identify expression differences between clusters: using mean expression level, or the ranking of the gene by its expression, or the proportions of cells that express the gene.

Our main objective in this section is to cover some of the methods that can be used to achieve this goal, and obtain a summary table of results

??? r-project-2 "Setup"

    ```r
    library(scater)
    library(scran)
    library(tidyverse)
    library(patchwork)
    ```

    - We will use the data set generated in the clustering session. This contains 7 samples from the Caron data set. For the purposes of these materials, in the interests of time, each sample has been downsampled to only contain 500 cells.
    - read single cell object

    ```r
    sce <- readRDS("R_objects/Caron_clustered.500.rds")
    ```
    ```r
    rownames(sce)[11:20]
    ```
    ```r
    all(sce$k.25_cluster.fun.leiden == sce$label)
    ```

    - To remind ourselves, we can visualise the clusters on a UMAP:
    ```r
    plotReducedDim(sce, 
                   dimred = "UMAP_corrected",
                   colour_by = "label", 
                   text_by = "label")
    ```
    <center>![image](../r_images/70-clustermarker-UMAP-corrected.png){width="400"}</center>

    - Our objective is to identify genes that distinguish these clusters from one another - “cluster marker genes”. Intuitively we hope that the clusters relate to specific cell populations, and therefore we are trying to find genes that will allow us to identify the cell types for each cluster.

    - For example genes such as the “CST3” gene, which is a known monocyte marker:

    ```r
    plotReducedDim(sce, 
                   dimred = "UMAP_corrected",
                   colour_by = "CST3", 
                   text_by = "label", 
                   by_exprs_values = "reconstructed",
                   add_legend = FALSE)
    ```
    <center>![image](../r_images/71-clustermarker-UMAPcorrected-2.png){width="400"}</center>


## Identifying cluster marker genes

Although we have defined our clusters based on the batch-corrected expression values, these should not be used for for gene-based analyses like marker gene detection. Instead, we should **use the uncorrected (normalised) expression values for differential expression between clusters**. This is because data integration algorithms bring cells together based on their overall gene expression, but for each gene individually the data transformation may introduce artificial agreement between batches, which is not ideal for gene-level differential analysis. Furthermore, the severity of these biases is dependent on the parameters used for data integration (such as the number of nearest neighbours in the `fastMNN()`` method).

Valid assays to use in gene based differential analysis tests are the normalised counts obtained from the deconvolution method (using `scuttle::computePooledFactors() + scuttle::logNormCounts()`) or from the variance stabilising transformation method (using `sctransform::vst(). In our SCE object, we have the normalised counts in the “logcounts” assay, which we can access with assay(`sce, "logcounts"` (or using the shortcut `logcounts(sce)`).

