# 9. Differential expression analysis - Pt.1

When we have data for multiple conditions (e.g. control vs treatment, healthy vs diseased, wild-type vs mutant, etc.) and multiple biological replicates for each of them, we can investigate which genes are differentially expressed between conditions. This is similar to what is done in more conventional bulk RNA-seq analysis, but in this case we have the added benefit of being able to do our analysis focusing on particular cell clusters/types (which would not be possible to do in a bulk experiment, unless we had done cell-sorting before the sequencing).

Differential analyses of multi-condition scRNA-seq experiments can be broadly split into two categories:

- **differential** expression (DE) tests for changes in expression between conditions for cells of the same ‘type’ that are present in both conditions.
- **differential** abundance (DA) tests for changes in the composition of cell types (or states, etc.) between conditions.
In this section we will cover the first of these, i.e. how to do a differential expression analysis for a simple design with a single factor (variable) with two levels (conditions). In our case this corresponds to the comparison between cells from healthy (PBMMC) versus leukemia (ETV6-RUNX1) donors.

There are three main steps for a bulk analysis:

- Creating the pseudo-bulk samples
- Filter low-count cells/samples
- Run the differential analysis


??? r-project-2 "Setup"

    ```r
    # load packages
    library(scater)
    library(scran)
    library(batchelor)
    library(bluster)
    library(edgeR)
    library(tidyverse)
    
    # load the SCE object
    sce <- readRDS("R_objects/Caron_clustered.PBMMCandETV6RUNX1.rds")
    
    # plot UMAP done on the batch-corrected data
    plotReducedDim(sce, dimred = "UMAP_corrected", 
                   colour_by = "label", 
                   text_by = "label")
    ```