
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/scRNA-seq-data-analysis/episodes/1/">
      
      
        <link rel="prev" href="../0/">
      
      
        <link rel="next" href="../2/">
      
      
      <link rel="icon" href="../../theme_images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>1. Alignment and feature counting with Cell Ranger - Analysis of single cell RNA-seq data</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#1-alignment-and-feature-counting-with-cell-ranger" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
<!-- Custom Banner -->
<div class="banner">
    <p style="color:rgb(250, 246, 246); display:inline;font-size: 160%;">This repository is in the early stages of development (Beta Version)</p>
</div>

<!-- Original Header -->


  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Analysis of single cell RNA-seq data" class="md-header__button md-logo" aria-label="Analysis of single cell RNA-seq data" data-md-component="logo">
      
  <img src="../../theme_images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Analysis of single cell RNA-seq data
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1. Alignment and feature counting with Cell Ranger
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/scRNA-seq-data-analysis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/scRNA-seq-data-analysis
  </div>
</a>
      </div>
    
  </nav>
  
</header>

    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Analysis of single cell RNA-seq data" class="md-nav__button md-logo" aria-label="Analysis of single cell RNA-seq data" data-md-component="logo">
      
  <img src="../../theme_images/nesi_ga.png" alt="logo">

    </a>
    Analysis of single cell RNA-seq data
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/scRNA-seq-data-analysis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/scRNA-seq-data-analysis
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Episodes
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Episodes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Workflow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    1. Alignment and feature counting with Cell Ranger
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    1. Alignment and feature counting with Cell Ranger
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#index-the-reference-genome" class="md-nav__link">
    <span class="md-ellipsis">
      Index the reference genome
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cell-ranger-count" class="md-nav__link">
    <span class="md-ellipsis">
      Cell Ranger count
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Cell Ranger count">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparing-the-raw-fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing the raw fastq files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-cellranger-count-note-this-takes-time-" class="md-nav__link">
    <span class="md-ellipsis">
      Running cellranger count ( NOTE: this takes time ---)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2. QC and Exploratory Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3. Normalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4. sctransform: Variance Stabilising Transformation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5. Feature Selection and Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6. Batch correction and data set integration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7. Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    8. Identification of cluster marker genes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../9.1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    9. Differential expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../9.2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    10. Differential Abundance
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#index-the-reference-genome" class="md-nav__link">
    <span class="md-ellipsis">
      Index the reference genome
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cell-ranger-count" class="md-nav__link">
    <span class="md-ellipsis">
      Cell Ranger count
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Cell Ranger count">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparing-the-raw-fastq-files" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing the raw fastq files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-cellranger-count-note-this-takes-time-" class="md-nav__link">
    <span class="md-ellipsis">
      Running cellranger count ( NOTE: this takes time ---)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="1-alignment-and-feature-counting-with-cell-ranger">1. Alignment and feature counting with Cell Ranger<a class="headerlink" href="#1-alignment-and-feature-counting-with-cell-ranger" title="Permanent link">&para;</a></h1>
<div class="admonition quote">
<p>The first step in the analysis of single cell RNAseq data is to align the sequenced reads against a genomic reference and then use a transcriptome annotation to generate read counts for each feature of interest. Typically for scRNAseq the features of interest are genes. </p>
<p>Due to time constraints, this will not be covered in today's workshop. We will begin this workshop with the filtered feature count output files generated by the Cell Ranger count function. </p>
<p>All the information relating to alignment and feature counting is contained below for you to read in your own time.</p>
</div>
<details class="file-code">
<summary>Indexing the reference genome and running CellRanger count</summary>
<p>There are a variety of tools for doing alighment and feature counting and your choice will depend in part on the method by which the library was generated. For data generated using the 10x-Chromium method data the most common approach is to use the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> tool provided by 10x. This not only carries out the alignment and feature counting, but will also:</p>
<ul>
<li>Call cells, i.e. filter the raw matrix to remove droplets that do not contain cells</li>
<li>Generate a very useful report in html format, which will provide some QC metrics and an initial look at the data</li>
<li>Generate a “cloupe” file, which can be opened using the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">10x Loupe Browser</a> software to further explore the data.</li>
</ul>
<p>Cell Ranger is computationally very intensive, you will not be able to run it on a laptop or standard desktop computer. You will need access to, for example, a high performance computing (HPC) cluster, a server or other cloud-based computational resource with sufficient power - talk to your local IT support.</p>
<p>Alternative methods include:</p>
<ul>
<li><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STAR solo</a> - this tool is built into the general purpose STAR aligner (Cell Ranger actually uses STAR under the hood). This will generate outputs very similar to Cell Ranger minus the cloupe file and the QC report. The advantage over Cell Ranger is that it is much less computationally intensive and will run with lower memory requirements in a shorter time.</li>
<li><a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin</a> - This tool is based on the popular Salmon tool for bulk RNAseq feature counting. Alevin supports both 10x-Chromium and Drop-seq derived data.</li>
</ul>
<p>For the purposes of this course, seeing as we are working with 10x-Chromium derived data, we will use Cell Ranger. As detailed instructions are available on the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger pages</a> of the 10x website, this chapter will not be comprehensive in terms of all options, but should provide a brief overview.</p>
<h2 id="index-the-reference-genome">Index the reference genome<a class="headerlink" href="#index-the-reference-genome" title="Permanent link">&para;</a></h2>
<p>To index our genome, we can use the cellranger mkref tool with the following arguments (replacing {...} with the relevant piece of information):</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>cellranger<span class="w"> </span>mkref<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w">  </span>--fasta<span class="o">={</span>GENOME<span class="w"> </span>FASTA<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w">  </span>--genes<span class="o">={</span>ANNOTATION<span class="w"> </span>GTF<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="w">  </span>--genome<span class="o">={</span>OUTPUT<span class="w"> </span>FOLDER<span class="w"> </span>FOR<span class="w"> </span>INDEX<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="w">  </span>--nthreads<span class="o">={</span>CPUS<span class="o">}</span>
</code></pre></div>
where:</p>
<div class="admonition info">
<ul>
<li><code>GENOME FASTA</code> is a file containing the reference genome in FASTA format.</li>
<li><code>ANNOTATION GTF</code> is a file containing the transcript annotation file in GTF format.</li>
<li><code>OUTPUT FOLDER FOR INDEX</code>is a name for the output folder containing the new reference package (you do not need to create this folder).</li>
<li><code>CPUS</code> - Is the number of CPUs we would like CellRanger to use. The more CPUs CellRanger can use, the faster the job (up to a point).</li>
</ul>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="nb">cd</span><span class="w"> </span>Data/references
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>module<span class="w"> </span>purge<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>module<span class="w"> </span>load<span class="w"> </span>CellRanger/7.1.0
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>cellranger<span class="w"> </span>mkref<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="w">  </span>--fasta<span class="o">=</span>Homo_sapiens.GRCh38.dna.chromosome.21.fa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="w">  </span>--genes<span class="o">=</span>gencode.v41.primary_assembly.annotation.chr21.gtf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="w">  </span>--genome<span class="o">=</span>cellranger_index<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="w">  </span>--nthreads<span class="w"> </span><span class="m">4</span>
</code></pre></div>
<details class="circle-check">
<summary>output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>[&#39;/scale_wlg_persistent/filesets/opt_nesi/CS400_centos7_bdw/CellRanger/7.1.0/bin/rna/mkref&#39;, &#39;--fasta=Homo_sapiens.GRCh38.dna.chromosome.21.fa&#39;, &#39;--genes=gencode.v41.primary_assembly.annotation.chr21.gtf&#39;, &#39;--genome=cellranger_index&#39;, &#39;--nthreads=7&#39;]
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>Creating new reference folder at /scale_wlg_persistent/filesets/project/nesi02659/sc-rna/Data/references/cellranger_index
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>...done
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>Writing genome FASTA file into reference folder...
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>...done
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>Indexing genome FASTA file...
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>...done
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>Writing genes GTF file into reference folder...
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>...done
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>Generating STAR genome index (may take over 8 core hours for a 3Gb genome)...
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>Aug 05 19:27:22 ..... started STAR run
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>Aug 05 19:27:22 ... starting to generate Genome files
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>Aug 05 19:27:23 ... starting to sort Suffix Array. This may take a long time...
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a>Aug 05 19:27:24 ... sorting Suffix Array chunks and saving them to disk...
<a id="__codelineno-2-19" name="__codelineno-2-19" href="#__codelineno-2-19"></a>Aug 05 19:27:52 ... loading chunks from disk, packing SA...
<a id="__codelineno-2-20" name="__codelineno-2-20" href="#__codelineno-2-20"></a>Aug 05 19:27:53 ... finished generating suffix array
<a id="__codelineno-2-21" name="__codelineno-2-21" href="#__codelineno-2-21"></a>Aug 05 19:27:53 ... generating Suffix Array index
<a id="__codelineno-2-22" name="__codelineno-2-22" href="#__codelineno-2-22"></a>Aug 05 19:27:55 ... completed Suffix Array index
<a id="__codelineno-2-23" name="__codelineno-2-23" href="#__codelineno-2-23"></a>Aug 05 19:27:55 ..... processing annotations GTF
<a id="__codelineno-2-24" name="__codelineno-2-24" href="#__codelineno-2-24"></a>Aug 05 19:27:55 ..... inserting junctions into the genome indices
<a id="__codelineno-2-25" name="__codelineno-2-25" href="#__codelineno-2-25"></a>Aug 05 19:27:58 ... writing Genome to disk ...
<a id="__codelineno-2-26" name="__codelineno-2-26" href="#__codelineno-2-26"></a>Aug 05 19:27:58 ... writing Suffix Array to disk ...
<a id="__codelineno-2-27" name="__codelineno-2-27" href="#__codelineno-2-27"></a>Aug 05 19:27:58 ... writing SAindex to disk
<a id="__codelineno-2-28" name="__codelineno-2-28" href="#__codelineno-2-28"></a>Aug 05 19:27:58 ..... finished successfully
<a id="__codelineno-2-29" name="__codelineno-2-29" href="#__codelineno-2-29"></a>...done.
<a id="__codelineno-2-30" name="__codelineno-2-30" href="#__codelineno-2-30"></a>
<a id="__codelineno-2-31" name="__codelineno-2-31" href="#__codelineno-2-31"></a>Writing genome metadata JSON file into reference folder...
<a id="__codelineno-2-32" name="__codelineno-2-32" href="#__codelineno-2-32"></a>Computing hash of genome FASTA file...
<a id="__codelineno-2-33" name="__codelineno-2-33" href="#__codelineno-2-33"></a>...done
<a id="__codelineno-2-34" name="__codelineno-2-34" href="#__codelineno-2-34"></a>
<a id="__codelineno-2-35" name="__codelineno-2-35" href="#__codelineno-2-35"></a>Computing hash of genes GTF file...
<a id="__codelineno-2-36" name="__codelineno-2-36" href="#__codelineno-2-36"></a>...done
<a id="__codelineno-2-37" name="__codelineno-2-37" href="#__codelineno-2-37"></a>
<a id="__codelineno-2-38" name="__codelineno-2-38" href="#__codelineno-2-38"></a>...done
<a id="__codelineno-2-39" name="__codelineno-2-39" href="#__codelineno-2-39"></a>
<a id="__codelineno-2-40" name="__codelineno-2-40" href="#__codelineno-2-40"></a>&gt;&gt;&gt; Reference successfully created! &lt;&lt;&lt;
<a id="__codelineno-2-41" name="__codelineno-2-41" href="#__codelineno-2-41"></a>
<a id="__codelineno-2-42" name="__codelineno-2-42" href="#__codelineno-2-42"></a>You can now specify this reference on the command line:
<a id="__codelineno-2-43" name="__codelineno-2-43" href="#__codelineno-2-43"></a>cellranger --transcriptome=/scale_wlg_persistent/filesets/project/nesi02659/sc-rna/Data/references/cellranger_index ...
</code></pre></div>
</details>
</div>
<h2 id="cell-ranger-count">Cell Ranger <code>count</code><a class="headerlink" href="#cell-ranger-count" title="Permanent link">&para;</a></h2>
<h3 id="preparing-the-raw-fastq-files">Preparing the raw fastq files<a class="headerlink" href="#preparing-the-raw-fastq-files" title="Permanent link">&para;</a></h3>
<p>To run the Cell Ranger count function, the fastq files for all samples to be processed should be placed in a single directory. Cell Ranger will be run separately on each sample. You will need to provide the software with the name of the sample to be processed (e.g. SRR9264343). Cell Ranger expects the file names of the fastq files to follow a specific convention so that it can correctly identify the files for the specific sample. If the files names of your fastqs do not match this convention you will need to rename them.</p>
<p>The convention is the default convention used by <code>bcl2fastq</code> (the tool that converts the raw sequencing data in <code>bcl</code> format into <code>fastq</code> files):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>&lt;SampleName&gt;_S&lt;SampleNumber&gt;_L00&lt;Lane&gt;_&lt;Read&gt;_001.fastq.gz
</code></pre></div>
<div class="admonition quote">
<ul>
<li><code>&lt;SampleName&gt;</code> - An identifier for the sample, this is what Cell Ranger uses to determine which fastq files to combine into a single sample.</li>
<li><code>&lt;SampleNumber&gt;</code> - This is the sample number based on the order that samples were listed in the sample sheet used when running bcl2fastq. This is not important for Cell Ranger, other than it indicates the end of the Sample Name, you can set all your samples to S1.</li>
<li><code>&lt;Lane&gt;</code> - The lane number. If your sequencing was run across multiple lanes, then you may have multiple sets of fastqs for a single sample with different lane numbers.</li>
<li><code>&lt;Read&gt;</code> - The read type: R1 for Read 1, R2 for Read 2, and index reads are I1 and I2.
001 - The last segment is always 001.</li>
</ul>
</div>
<p>e.g. for a single sample in the Caron data set we have:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="w">    </span>SRR9264343_S0_L001_I1_001.fastq.gz
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="w">    </span>SRR9264343_S0_L001_R1_001.fastq.gz
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="w">    </span>SRR9264343_S0_L001_R2_001.fastq.gz
</code></pre></div>
<h3 id="running-cellranger-count-note-this-takes-time-">Running <code>cellranger count</code> ( NOTE: this takes time ---)<a class="headerlink" href="#running-cellranger-count-note-this-takes-time-" title="Permanent link">&para;</a></h3>
<p>The minimum information require to run cellranger count is:</p>
<div class="admonition quote">
<ul>
<li><code>--id</code> - A sample ID. This is used for naming the outputs</li>
<li><code>--transcriptome</code>- the directory containing the Cell Ranger reference</li>
<li>
<p><code>--fastqs</code> - the directory containing the fastq files
This will process all fastq files in the --fastqs directory into a single sample. If you have multiple samples in a single directory then you need to add:</p>
</li>
<li>
<p><code>--sample</code> - the SampleName from the fastq files as above.
In addition, Cell Ranger is very computationally intensive, you will usually be wanting to run it on a high performance cluster or server. It will greedily attempt to use all resources it can find available, and so it is advisable to set limits to the resources it will use:</p>
</li>
<li>
<p><code>--localcores</code> - the number of processors Cell Ranger should use</p>
</li>
<li><code>--localmem</code> - the amount of memory, in Gigabytes, Cell Ranger should use.</li>
</ul>
<p>A complete command for processing on of the Caron samples might be:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>cellranger<span class="w"> </span>count<span class="w"> </span>--id<span class="o">={</span>OUTPUT_SAMPLE_NAME<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="w">                 </span>--transcriptome<span class="o">={</span>DIRECTORY_WITH_REFERENCE<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="w">                 </span>--fastqs<span class="o">={</span>DIRECTORY_WITH_FASTQ_FILES<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="w">                 </span>--sample<span class="o">={</span>NAME_OF_SAMPLE_IN_FASTQ_FILES<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="w">                 </span>--localcores<span class="o">={</span>NUMBER_OF_CPUS<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a><span class="w">                 </span>--localmem<span class="o">={</span>RAM_MEMORY<span class="o">}</span>
</code></pre></div>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>cellranger<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="w">  </span>--id<span class="o">=</span>ETV6_RUNX1_rep1<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="w">   </span>--transcriptome<span class="o">=</span>Data/references/cellranger_index<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="w">   </span>--fastqs<span class="o">=</span>Data/fastq/<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="w">   </span>--sample<span class="o">=</span>SRR9264343<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="w">   </span>--localcores<span class="o">=</span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="w">   </span>--localmem<span class="o">=</span><span class="m">24</span>
</code></pre></div>
<details class="file-code">
<summary>Slurm-script - This will run for about 9 minutes</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="ch">#!/bin/bash -e </span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="c1">#SBATCH --account nesi02659</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="c1">#SBATCH --job-name cell-ranger-count</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="c1">#SBATCH --cpus-per-task 16</span>
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="c1">#SBATCH --mem 24G</span>
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="c1">#SBATCH --time 00:20:00 </span>
<a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a><span class="c1">#SBATCH --output slurmlog/%j.out</span>
<a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a>
<a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a>module<span class="w"> </span>purge
<a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a>module<span class="w"> </span>load<span class="w"> </span>CellRanger/7.1.0
<a id="__codelineno-7-12" name="__codelineno-7-12" href="#__codelineno-7-12"></a>
<a id="__codelineno-7-13" name="__codelineno-7-13" href="#__codelineno-7-13"></a>cellranger<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-14" name="__codelineno-7-14" href="#__codelineno-7-14"></a><span class="w">            </span>--id<span class="o">=</span>ETV6_RUNX1_rep1<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-15" name="__codelineno-7-15" href="#__codelineno-7-15"></a><span class="w">            </span>--transcriptome<span class="o">=</span>Data/references/cellranger_index<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-16" name="__codelineno-7-16" href="#__codelineno-7-16"></a><span class="w">            </span>--fastqs<span class="o">=</span>Data/fastq/<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-17" name="__codelineno-7-17" href="#__codelineno-7-17"></a><span class="w">            </span>--sample<span class="o">=</span>SRR9264343<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-18" name="__codelineno-7-18" href="#__codelineno-7-18"></a><span class="w">            </span>--localcores<span class="o">=</span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-19" name="__codelineno-7-19" href="#__codelineno-7-19"></a><span class="w">            </span>--localmem<span class="o">=</span><span class="m">22</span>
</code></pre></div>
</details>
<details class="circle-check">
<summary><code>ls -F ETV6_RUNX1_rep1/outs/</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>analysis/<span class="w">                    </span>filtered_feature_bc_matrix.h5<span class="w">  </span>possorted_genome_bam.bam<span class="w">      </span>raw_feature_bc_matrix.h5
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>cloupe.cloupe<span class="w">                </span>metrics_summary.csv<span class="w">            </span>possorted_genome_bam.bam.bai<span class="w">  </span>web_summary.html
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>filtered_feature_bc_matrix/<span class="w">  </span>molecule_info.h5<span class="w">               </span>raw_feature_bc_matrix/
</code></pre></div>
</details>
</div>
</details>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../0/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Workflow">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Workflow
              </div>
            </div>
          </a>
        
        
          
          <a href="../2/" class="md-footer__link md-footer__link--next" aria-label="Next: 2. QC and Exploratory Analysis">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                2. QC and Exploratory Analysis
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Analysis of single cell RNA-seq data workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>