
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/scRNA-seq-data-analysis/episodes/3/">
      
      
        <link rel="prev" href="../2/">
      
      
        <link rel="next" href="../4/">
      
      
      <link rel="icon" href="../../theme_images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>3. Normalisation - Analysis of single cell RNA-seq data</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#3-normalisation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
<!-- Custom Banner -->
<div class="banner">
    <p style="color:rgb(250, 246, 246); display:inline;font-size: 160%;">This repository is in the early stages of development (Beta Version)</p>
</div>

<!-- Original Header -->


  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Analysis of single cell RNA-seq data" class="md-header__button md-logo" aria-label="Analysis of single cell RNA-seq data" data-md-component="logo">
      
  <img src="../../theme_images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Analysis of single cell RNA-seq data
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              3. Normalisation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/scRNA-seq-data-analysis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/scRNA-seq-data-analysis
  </div>
</a>
      </div>
    
  </nav>
  
</header>

    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Analysis of single cell RNA-seq data" class="md-nav__button md-logo" aria-label="Analysis of single cell RNA-seq data" data-md-component="logo">
      
  <img src="../../theme_images/nesi_ga.png" alt="logo">

    </a>
    Analysis of single cell RNA-seq data
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/scRNA-seq-data-analysis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/scRNA-seq-data-analysis
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Episodes
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Episodes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Workflow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1. Alignment and feature counting with Cell Ranger
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2. QC and Exploratory Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    3. Normalisation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    3. Normalisation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#load-data-object" class="md-nav__link">
    <span class="md-ellipsis">
      Load data object
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#why-normalise" class="md-nav__link">
    <span class="md-ellipsis">
      Why normalise ?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#normalization-strategies" class="md-nav__link">
    <span class="md-ellipsis">
      Normalization strategies
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deconvolution" class="md-nav__link">
    <span class="md-ellipsis">
      Deconvolution
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Deconvolution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cluster-cells" class="md-nav__link">
    <span class="md-ellipsis">
      Cluster Cells
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute-size-factors" class="md-nav__link">
    <span class="md-ellipsis">
      Compute size factors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#apply-size-factors" class="md-nav__link">
    <span class="md-ellipsis">
      Apply size factors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#explore-the-effect-of-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      Explore the effect of normalisation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#save-the-normalised-object" class="md-nav__link">
    <span class="md-ellipsis">
      Save the normalised object
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4. sctransform: Variance Stabilising Transformation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5. Feature Selection and Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6. Batch correction and data set integration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7. Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    8. Identification of cluster marker genes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../9.1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    9. Differential expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../9.2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    10. Differential Abundance
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#load-data-object" class="md-nav__link">
    <span class="md-ellipsis">
      Load data object
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#why-normalise" class="md-nav__link">
    <span class="md-ellipsis">
      Why normalise ?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#normalization-strategies" class="md-nav__link">
    <span class="md-ellipsis">
      Normalization strategies
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deconvolution" class="md-nav__link">
    <span class="md-ellipsis">
      Deconvolution
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Deconvolution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cluster-cells" class="md-nav__link">
    <span class="md-ellipsis">
      Cluster Cells
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute-size-factors" class="md-nav__link">
    <span class="md-ellipsis">
      Compute size factors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#apply-size-factors" class="md-nav__link">
    <span class="md-ellipsis">
      Apply size factors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#explore-the-effect-of-normalisation" class="md-nav__link">
    <span class="md-ellipsis">
      Explore the effect of normalisation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#save-the-normalised-object" class="md-nav__link">
    <span class="md-ellipsis">
      Save the normalised object
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="3-normalisation">3. Normalisation<a class="headerlink" href="#3-normalisation" title="Permanent link">&para;</a></h1>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nf">library</span><span class="p">(</span><span class="n">scater</span><span class="p">)</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="nf">library</span><span class="p">(</span><span class="n">scran</span><span class="p">)</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="nf">library</span><span class="p">(</span><span class="n">sctransform</span><span class="p">)</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="nf">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="nf">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="n">bpp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">MulticoreParam</span><span class="p">(</span><span class="m">4</span><span class="p">)</span>
</code></pre></div>
</div>
<h2 id="load-data-object">Load data object<a class="headerlink" href="#load-data-object" title="Permanent link">&para;</a></h2>
<p>For the purposes of this demonstration we have generated a smaller data set in which there are only 500 cells per sample. This is so that the code can be run in a reasonable amount of time during the live teaching session. The data were first QC’d and filtered as described in the QC and exploratory analysis session. After this 500 cells were selected at random from each sample.</p>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="n">sce</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;R_objects/Caron_filtered.500.rds&quot;</span><span class="p">)</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="n">sce</span>
</code></pre></div>
<details class="circle-check">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>class: SingleCellExperiment 
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>dim: 29346 5500 
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>metadata(1): Samples
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>assays(1): counts
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>rownames(29346): ENSG00000243485 ENSG00000238009 ... ENSG00000275063 ENSG00000278817
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>rowData names(4): ID Symbol Type Chromosome
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>colnames(5500): 1_CGACTTCGTCCAGTTA-1 1_AGAATAGCATACGCTA-1 ... 8_AGCGTCGAGATGTAAC-1 8_CATGACAAGATAGGAG-1
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>colData names(10): Sample Barcode ... subsets_Mito_percent total
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>reducedDimNames(0):
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>mainExpName: NULL
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>altExpNames(0):
</code></pre></div>
</details>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="nf">table</span><span class="p">(</span><span class="n">sce</span><span class="o">$</span><span class="n">SampleName</span><span class="p">)</span>
</code></pre></div>
</div>
<h2 id="why-normalise">Why normalise ?<a class="headerlink" href="#why-normalise" title="Permanent link">&para;</a></h2>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="n">oneSamTab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">colData</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="w">  </span><span class="nf">as.data.frame</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="w">  </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">SampleName</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;PBMMC_1&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="w">  </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">select</span><span class="p">(</span><span class="n">SampleName</span><span class="p">,</span><span class="n">Barcode</span><span class="p">,</span><span class="w"> </span><span class="n">sum</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cell_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">n</span><span class="p">())</span>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="n">p_before_nom</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">oneSamTab</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_num</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">sum</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a><span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;identity&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a><span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Cell Index&#39;</span><span class="p">,</span>
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a><span class="w">        </span><span class="n">y</span><span class="o">=</span><span class="s">&#39;Cell UMI counts&#39;</span><span class="p">,</span>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a><span class="w">        </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;PBMMC_1: Before Normalization&quot;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a><span class="w">  </span><span class="nf">theme_classic</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a><span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a><span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;red&#39;</span><span class="p">)</span>
<a id="__codelineno-4-15" name="__codelineno-4-15" href="#__codelineno-4-15"></a><span class="w">  </span><span class="p">)</span>
<a id="__codelineno-4-16" name="__codelineno-4-16" href="#__codelineno-4-16"></a>
<a id="__codelineno-4-17" name="__codelineno-4-17" href="#__codelineno-4-17"></a><span class="n">p_before_nom</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/15-normalise-why-normalise.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/15-normalise-why-normalise.png" width="600" /></a></p>
<div class="admonition rectangle-list">
<ul>
<li>
<p>Above plot shows the UMI counts/cell (transcript molecules) for 500 cell from the PBMMC_1 sample</p>
</li>
<li>
<p>UMI counts fluctuate</p>
</li>
<li>
<p>We derive biological insights downstream by comparing cells against each other.</p>
</li>
<li>
<p>But the UMI counts differences makes it harder to compare cells.</p>
</li>
<li>
<p>Why total transcript molecules (UMI counts) detected between cells differ?</p>
<ul>
<li>
<p>Biological:</p>
<ul>
<li>Cell sub type differences, like size and transcription activity etc.</li>
</ul>
</li>
<li>
<p>Technical: scRNA data is inherently noisy</p>
<ul>
<li>Low mRNA content per cell</li>
<li>cell-to-cell differences in mRNA capture efficiency</li>
<li>Variable sequencing depth</li>
<li>PCR amplification efficiency</li>
</ul>
</li>
</ul>
</li>
<li>
<p>A normalization technique makes the UMI counts distribution uniform, so that each cell has similar counts.</p>
</li>
<li>
<p>Normalization is a critical step that corrects cell-to-cell technical differences.</p>
</li>
<li>
<p>By normalizing, downstream comparisons of relative expression between cells are valid.</p>
</li>
</ul>
</div>
</div>
<h2 id="normalization-strategies">Normalization strategies<a class="headerlink" href="#normalization-strategies" title="Permanent link">&para;</a></h2>
<div class="admonition rectangle-list">
<p class="admonition-title">The sparse nature of scRNA data makes normalization difficult, unlike bulk RNAseq data</p>
<p>Broadly two classes:</p>
<ul>
<li>
<p>Spike-in methods</p>
<ul>
<li>Uses spike-in controls for normalisation</li>
<li>Not available for droplet based scRNA techniques like 10x.</li>
</ul>
</li>
<li>
<p>Non-spike-in methods: </p>
<ul>
<li>Using available counts data for normalization</li>
<li>DEseq2</li>
<li>edgeR-TMM</li>
<li>Library size normalization</li>
<li>deconvolution</li>
<li>sctransform</li>
</ul>
</li>
</ul>
<p>Typical normalization has two steps:</p>
<ul>
<li>
<p>Scaling</p>
<ul>
<li>Estimate size or scaling or normalization factor: computation of a cell-specific ‘scaling’ or ‘size’ factor or “normalization factor” that represents the relative bias in that cell and division of all counts for the cell by that factor to remove that bias. Assumption: any cell specific bias will affect genes the same way.</li>
<li>Scale the data by dividing the count for each gene with the appropriate size factor for that cell        </li>
</ul>
</li>
<li>
<p>Transformation</p>
<ul>
<li>log2</li>
<li>Square root transformation</li>
<li>Pearson residuals (eg. sctransform)</li>
</ul>
</li>
</ul>
<p>Scaling methods typically generate normalised counts-per-million (CPM) or transcripts-per-million (TPM) values that address the effect of sequencing depth. These values however typically have a variance that increases with their mean (heteroscedasticity) while most statistical methods assume a stable variance, which does not vary with the mean (homoscedasticity). A widely used ‘variance stabilising transformation’ is the log transformation (often log2). This works well for highly expressed genes (as in bulk RNA-seq) but less so for sparse scRNA-seq data.</p>
<p><center><a class="glightbox" href="../../r_images/normalisation_strategies.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/normalisation_strategies.png" /></a></center>
<small>* DEseq, edgeR-TMM and Library size normalization initially developed for bulk RNAseq * Applying these methods on scRNAseq data systematically under or over estimate size factors. i.e systematically deviate from true size factors. * This deviation is the result of removing zeroes prior to normalization. * Therefore other normalization methods specific to scRNAseq data like deconvolution, sctransform etc. were proposed.</small></p>
</div>
<h2 id="deconvolution">Deconvolution<a class="headerlink" href="#deconvolution" title="Permanent link">&para;</a></h2>
<p>Because single-cell data tend to have a substantial number of low and zero counts, these bulk normalization methods may be problematic for single-cell data.</p>
<ul>
<li>Deconvolution aims to normalize expression values based on summed values from pools of cells.</li>
<li>Since cell summation results in fewer zeros, the ensuing normalization is less susceptible to errors than existing methods.</li>
<li>The estimated size factors are only relevant to the pools of cells, even though normalization accuracy has improved.</li>
<li>Each pool’s size factor is deconvolved into its constituent cells’ size factors.</li>
</ul>
<div class="admonition stairs">
<p class="admonition-title">The deconvolution method consists of several key steps:</p>
<p><small><strong>References:</strong> <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7</a></small></p>
<ul>
<li>Defining a pool of cells</li>
<li>Summing expression values across all cells in the pool</li>
<li>Normalizing the cell pool against an average reference, using the summed expression values</li>
<li>Repeating this for many different pools of cells to construct a linear system</li>
<li>Deconvolving the pool-based size factors to their cell-based counterparts </li>
</ul>
<p><center><a class="glightbox" href="../../r_images/normalisation-deconvolution-methods.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/normalisation-deconvolution-methods.png" width="600" /></a></center></p>
<p><small>Schematic of the deconvolution method. All cells in the data set are averaged to make a reference pseudo-cell. Expression values for cells in pool A are summed together and normalised against the reference to yield a pool-based size factor <em>θ</em><sub>A</sub> . This is equal to the sum of the cell-based factors <em>θ</em><sub>j</sub> for cells <em>j</em>=1–4 and can be used to formulate a linear equation. (For simplicity, the <em>t</em><sub>j</sub> term is assumed to be unity here.) Repeating this for multiple pools (e.g., pool B) leads to the construction of a linear system that can be solved to estimate <em>θ</em><sub>j</sub> for each cell <em>j</em></small></p>
<p>In order to avoid pooling cells with radically different transcriptomic profiles, the cells are first clustered based on gene expression. The pools are then formed exclusively with each cluster. Size factors are calculated within each cluster and are then scaled so they are comparable across clusters.</p>
</div>
<h3 id="cluster-cells">Cluster Cells<a class="headerlink" href="#cluster-cells" title="Permanent link">&para;</a></h3>
<p>The table below show the number and size of clusters found:</p>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="nf">set.seed</span><span class="p">(</span><span class="m">100</span><span class="p">)</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="n">clust</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">quickCluster</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="n">BPPARAM</span><span class="o">=</span><span class="n">bpp</span><span class="p">)</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="nf">table</span><span class="p">(</span><span class="n">clust</span><span class="p">)</span>
</code></pre></div>
</div>
<h3 id="compute-size-factors">Compute size factors<a class="headerlink" href="#compute-size-factors" title="Permanent link">&para;</a></h3>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">sce</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">computePooledFactors</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="w">             </span><span class="n">clusters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">clust</span><span class="p">,</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="w">             </span><span class="n">min.mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">,</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="w">             </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bpp</span><span class="p">)</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="n">deconv.sf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sizeFactors</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="nf">summary</span><span class="p">(</span><span class="n">deconv.sf</span><span class="p">)</span>
</code></pre></div>
<strong>Note:</strong> min.mean - A numeric scalar specifying the minimum (library size-adjusted) average count of genes to be used for normalization. This means large numbers of very lowly expressed genes will not bias the normalization.</p>
<p>Plot deconvolution size factors against library size factors:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="n">lib.sf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">librarySizeFactors</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="nf">data.frame</span><span class="p">(</span><span class="n">LibrarySizeFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lib.sf</span><span class="p">,</span><span class="w"> </span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="w">           </span><span class="n">DeconvolutionSizeFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">deconv.sf</span><span class="p">,</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="w">                 </span><span class="n">SampleGroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sce</span><span class="o">$</span><span class="n">SampleGroup</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="w">  </span><span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">LibrarySizeFactors</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">DeconvolutionSizeFactors</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="w">      </span><span class="nf">geom_point</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="n">SampleGroup</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="w">      </span><span class="nf">geom_abline</span><span class="p">(</span><span class="n">slope</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">intercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/16-normalisation-compute-size-factors.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/16-normalisation-compute-size-factors.png" width="600" /></a></p>
</div>
<h3 id="apply-size-factors">Apply size factors<a class="headerlink" href="#apply-size-factors" title="Permanent link">&para;</a></h3>
<p>For each cell, raw counts for genes are divided by the size factor for that cell and log-transformed so downstream analyses focus on genes with strong relative differences. We use <code>scater::logNormCounts()</code>.</p>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="n">sce</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">logNormCounts</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="nf">assayNames</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
</code></pre></div>
</div>
<h3 id="explore-the-effect-of-normalisation">Explore the effect of normalisation<a class="headerlink" href="#explore-the-effect-of-normalisation" title="Permanent link">&para;</a></h3>
<p>Normalised counts are much less variable across cells than raw counts</p>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="n">norm_counts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">logNormCounts</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="n">transform</span><span class="o">=</span><span class="s">&#39;none&#39;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="w">  </span><span class="nf">assay</span><span class="p">(</span><span class="s">&#39;normcounts&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="w">  </span><span class="nf">as.matrix</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="w">  </span><span class="nf">colSums</span><span class="p">()</span>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a><span class="n">norm_counts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span><span class="n">Barcode</span><span class="o">=</span><span class="nf">names</span><span class="p">(</span><span class="n">norm_counts</span><span class="p">),</span>
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="w">                      </span><span class="n">normCounts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">log2</span><span class="p">(</span><span class="n">norm_counts</span><span class="p">)</span>
<a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a><span class="w">                      </span><span class="p">)</span>
<a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a><span class="n">norm_counts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">inner_join</span><span class="p">(</span><span class="n">norm_counts</span><span class="p">,</span><span class="w"> </span><span class="n">oneSamTab</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s">&#39;Barcode&#39;</span><span class="p">)</span>
<a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a>
<a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a>
<a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a><span class="n">p_after_norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">norm_counts</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_num</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">normCounts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-9-12" name="__codelineno-9-12" href="#__codelineno-9-12"></a><span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;identity&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-9-13" name="__codelineno-9-13" href="#__codelineno-9-13"></a><span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Cell Index&#39;</span><span class="p">,</span>
<a id="__codelineno-9-14" name="__codelineno-9-14" href="#__codelineno-9-14"></a><span class="w">        </span><span class="n">y</span><span class="o">=</span><span class="s">&#39;Normalized Cell UMI counts&#39;</span><span class="p">,</span>
<a id="__codelineno-9-15" name="__codelineno-9-15" href="#__codelineno-9-15"></a><span class="w">        </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;PBMMC_1:After Normalization&quot;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-9-16" name="__codelineno-9-16" href="#__codelineno-9-16"></a><span class="w">  </span><span class="nf">theme_classic</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-9-17" name="__codelineno-9-17" href="#__codelineno-9-17"></a><span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<a id="__codelineno-9-18" name="__codelineno-9-18" href="#__codelineno-9-18"></a><span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;red&#39;</span><span class="p">)</span>
<a id="__codelineno-9-19" name="__codelineno-9-19" href="#__codelineno-9-19"></a><span class="w">  </span><span class="p">)</span>
<a id="__codelineno-9-20" name="__codelineno-9-20" href="#__codelineno-9-20"></a>
<a id="__codelineno-9-21" name="__codelineno-9-21" href="#__codelineno-9-21"></a><span class="n">p_before_nom</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p_after_norm</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/17-normalisation-explore-the-effect.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/17-normalisation-explore-the-effect.png" width="600" /></a></p>
<p>Let’s separate out the scaling normalisation from the log transformation</p>
<p>What do the un-normalised data look like if we log them?</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">p_before_norm_log</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">oneSamTab</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_num</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="nf">log2</span><span class="p">(</span><span class="n">sum</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;identity&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Cell Index&#39;</span><span class="p">,</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="w">        </span><span class="n">y</span><span class="o">=</span><span class="s">&#39;Cell UMI counts&#39;</span><span class="p">,</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="w">        </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Logged raw counts&quot;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a><span class="w">  </span><span class="nf">theme_classic</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<a id="__codelineno-10-7" name="__codelineno-10-7" href="#__codelineno-10-7"></a><span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<a id="__codelineno-10-8" name="__codelineno-10-8" href="#__codelineno-10-8"></a><span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;red&#39;</span><span class="p">)</span>
<a id="__codelineno-10-9" name="__codelineno-10-9" href="#__codelineno-10-9"></a><span class="w">  </span><span class="p">)</span>
<a id="__codelineno-10-10" name="__codelineno-10-10" href="#__codelineno-10-10"></a>
<a id="__codelineno-10-11" name="__codelineno-10-11" href="#__codelineno-10-11"></a><span class="n">p_before_norm_log</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p_after_norm</span>
</code></pre></div>
<p><a class="glightbox" href="../../r_images/18-normalisation-explore-the-effect-separate.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/18-normalisation-explore-the-effect-separate.png" width="600" /></a></p>
<p>Simply logging the sum of counts per cell reduces the variation a lot, but the scaling is required to do the job properly.</p>
<p>The log transformation is meant to reduce the correlation between mean and variance for genes - has this worked?</p>
<p>We can look at the relationship between the mean gene expression and variance for raw UMI counts, scaled counts and scaled, logged counts.</p>
<p>For raw counts:
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="c1"># mean and variance for raw counts</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="n">mean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowMeans</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;counts&quot;</span><span class="p">))</span>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="n">var</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowVars</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;counts&quot;</span><span class="p">))</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="c1"># Scatter plot</span>
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="nf">plot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">mean</span><span class="p">),</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">var</span><span class="p">),</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw data&quot;</span><span class="p">)</span>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="nf">abline</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">)</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/19-normalisation-mean-abline.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/19-normalisation-mean-abline.png" width="600" /></a></p>
<p>There is a strong linear relationship between mean and variance of UMI counts across genes.</p>
<p>For scaled counts:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="c1"># Mean and variance for scaled counts</span>
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="n">mean_scaled</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">logNormCounts</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="n">transform</span><span class="o">=</span><span class="s">&#39;none&#39;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="w">  </span><span class="nf">assay</span><span class="p">(</span><span class="s">&#39;normcounts&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="w">  </span><span class="nf">rowMeans</span><span class="p">()</span>
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="n">var_scaled</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">logNormCounts</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="n">transform</span><span class="o">=</span><span class="s">&#39;none&#39;</span><span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="w">  </span><span class="nf">assay</span><span class="p">(</span><span class="s">&#39;normcounts&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="w">  </span><span class="nf">rowVars</span><span class="p">()</span>
<a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a>
<a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a><span class="nf">plot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">mean_scaled</span><span class="p">),</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">var_scaled</span><span class="p">),</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;scaled data&#39;</span><span class="p">)</span>
<a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a><span class="nf">abline</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">)</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/20-normalisation-mean-scaled-counts.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/20-normalisation-mean-scaled-counts.png" width="600" /></a></p>
<p>The relationship is still there after scaling the counts.</p>
<p>For scaled, log transformed counts:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="c1"># Mean and variance for scaled, log transformed counts</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="n">mean_norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowMeans</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;logcounts&quot;</span><span class="p">))</span>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a><span class="n">var_norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowVars</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;logcounts&quot;</span><span class="p">))</span>
<a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a>
<a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a><span class="nf">plot</span><span class="p">(</span><span class="n">mean_norm</span><span class="p">,</span><span class="w"> </span><span class="n">var_norm</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;</span><span class="err">scaled, log transformed)</span>
<a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="nf">abline</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">)</span>
</code></pre></div>
<a class="glightbox" href="../../r_images/21-normalisation-relship-scaling-and-counts.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="image" src="../../r_images/21-normalisation-relship-scaling-and-counts.png" width="600" /></a></p>
<p>We see that the log transformation removes a large part of the relationship between mean and variance for gene expression values.</p>
</div>
<h2 id="save-the-normalised-object">Save the normalised object<a class="headerlink" href="#save-the-normalised-object" title="Permanent link">&para;</a></h2>
<p>Make sure to create the <code>results</code> directory first.</p>
<div class="admonition r-project">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="nf">saveRDS</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;results/caron_normalised.rds&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
<details class="clipboard-question">
<summary>Exercise</summary>
<p>Apply the deconvolution normalisation on a single sample: ETV6-RUNX1_1.</p>
<p>You will first load the a single cell experiment object containing the entire Caron data set. First it is necessary to select only cells that came from the sample ‘ETV6-RUNX1_1’. You can then apply the normalization by deconvolution by clustering the cells, computing size factors and using these to log-normalize the counts.</p>
</details>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../2/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 2. QC and Exploratory Analysis">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                2. QC and Exploratory Analysis
              </div>
            </div>
          </a>
        
        
          
          <a href="../4/" class="md-footer__link md-footer__link--next" aria-label="Next: 4. sctransform: Variance Stabilising Transformation">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                4. sctransform: Variance Stabilising Transformation
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Analysis of single cell RNA-seq data workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>