<center>
# Analysis of single cell RNA-seq data
</center>


!!! database "Data set"

    - Data used in this workshop is based on [CaronBourque2020](https://www.nature.com/articles/s41598-020-64929-x_) relating to pediatric leukemia, with four sample types, including:
        - pediatric Bone Marrow Mononuclear Cells (PBMMCs)
        - three tumour types: ETV6-RUNX1, HHD, PRE-T



!!! check-to-slot "Prerequisites"
 
    - [x] Inermediate level knowledge on R (programming language)
    - [x] Familiarity with terminal and basic linux commands
    - [x] Some knowledge on shell environment variables and `for` loops
    - [x] Ability to use a terminal based text editor such as `nano` 
        * [ ] This is not much of an issue as we are using JupyterHub which has a more friendlier text editor.   
    - [x] Intermediate level knowledge on Molecular Biology and Genetics

    **Recommended but not required**

    - [ ] Attend [Genomics Data Carpentry](https://datacarpentry.org/genomics-workshop/),  [RNA-Seq Data Analysis](https://genomicsaotearoa.github.io/RNA-seq-workshop/), [Introduction to R](https://genomicsaotearoa.github.io/Introduction-to-R/) and/or [Intermediate R](https://genomicsaotearoa.github.io/Intermediate-R/)

<br>

!!! book-atlas "References"

    - https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7
    - https://www.nature.com/articles/nbt.4091
    - https://arxiv.org/abs/physics/0512106
    - https://www.nature.com/articles/s41598-019-41695-z
    - https://www.nature.com/articles/s41598-019-41695-z#Fig3